logstash-core/src/main/java/org/logstash/config/ir/PipelineConfig.java

```diff
@@ -0,0 +1,180 @@
+import java.util.stream.Collectors;
+
+import static org.logstash.RubyUtil.RUBY;
+
+public final class PipelineConfig {
+
+    private static class LineToSource {
+        private final int startLine;
+        private final int endLine;
+        private final SourceWithMetadata source;
+
+        LineToSource(int startLine, int endLine, SourceWithMetadata source) {
+            this.startLine = startLine;
+            this.endLine = endLine;
+            this.source = source;
+        }
+
+        boolean includeLine(int lineNumber) {
+            return startLine <= lineNumber && lineNumber <= endLine;
+        }
+    }
+
+    private static final Logger logger = LogManager.getLogger(PipelineConfig.class);
+
+    private RubyClass source;
+    private String pipelineId;
+    private List<SourceWithMetadata> confParts;
+    private RubyObject settings;
+    private LocalDateTime readAt;
+    private String configHash;
+    private String configString;
+    private List<LineToSource> sourceRefs;
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    public PipelineConfig(RubyClass source, RubySymbol pipelineId, RubyObject uncastedConfigParts, RubyObject logstashSettings) {
+        IRubyObject uncasted = uncastedConfigParts.checkArrayType();
+        final RubyArray configParts = !uncasted.isNil() ?
+                (RubyArray) uncasted :
+                RubyArray.newArray(RUBY, uncastedConfigParts);
+
+        this.source = source;
+        this.pipelineId = pipelineId.toString();
+        SourceWithMetadata[] castedConfigParts = (SourceWithMetadata[]) configParts.toJava(SourceWithMetadata[].class);
+        List<SourceWithMetadata> confParts = Arrays.asList(castedConfigParts);
+        confParts.sort(Comparator.comparing(SourceWithMetadata::getProtocol)
+                .thenComparing(SourceWithMetadata::getId));
+        this.confParts = confParts;
+        this.settings = logstashSettings;
+        this.readAt = LocalDateTime.now();
+    }
+
+    public RubyClass getSource() {
+        return source;
+    }
+
+    public String getPipelineId() {
+        return pipelineId;
+    }
+
+    public List<SourceWithMetadata> getConfigParts() {
+        return confParts;
+    }
+
+    public LocalDateTime getReadAt() {
+        return readAt;
+    }
+
+    public RubyObject getSettings() {
+        return settings;
+    }
+
+    public String configHash() {
+        if (configHash == null) {
+            configHash = DigestUtils.sha1Hex(configString());
+        }
+        return configHash;
+    }
+
+    public String configString() {
+        this.configString = confParts.stream().map(SourceWithMetadata::getText).collect(Collectors.joining("\n"));
+        return this.configString;
+    }
+
+    public boolean isSystem() {
+        return this.settings.callMethod(RUBY.getCurrentContext(), "get_value",
+                                        RubyString.newString(RUBY, "pipeline.system"))
+                .toJava(Boolean.class);
```

Comment:
`isTrue()` reads better ... e.g. a potential `nil` might not convert to a Java boolean but still has false-y semantics
```suggestion
                .isTrue();
```